### PR TITLE
Add unique constraints

### PR DESCRIPTION
**Pull request by gidsg**
### Context
Previously we had no constraints within the database to prevent the creation of duplicate entries

### Changes proposed in this pull request
add unique index on `hash_value entry_type timestamp spina_register_id` 

### Guidance to review
`rake db:migrate` should add index to DB.

### CODE REVIEW

**Comment by arnau** [comment]
@michaelabenyohai @karlbaker02 @danmitchell- this is something everyone should validate it's not going to backfire

File: db/migrate/20180130160640_add_unique_constraint_to_entries.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddUniqueConstraintToEntries < ActiveRecord::Migration[5.1]
+  def change
+    add_index(:entries, %i[hash_value entry_type timestamp spina_register_id], unique: true, name: 'unique_entry_index')
```
Comment (karlbaker02) [inline db/migrate/20180130160640_add_unique_constraint_to_entries.rb]: Technically you could have multiple entries created at the same time for the same register and which point at the same object. The index should instead be on `[hash_value entry_type entry_number spina_register_id]`, as `entry_number` is guaranteed to be unique for each `entry_type`.

File: db/migrate/20180130160640_add_unique_constraint_to_entries.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 class AddUniqueConstraintToEntries < ActiveRecord::Migration[5.1]
   def change
-    add_index(:entries, %i[hash_value entry_type timestamp spina_register_id], unique: true, name: 'unique_entry_index')
+    add_index(:entries, %i[hash_value entry_type entry_number spina_register_id], unique: true, name: 'unique_entry_index')
```
Comment (gidsg) [inline db/migrate/20180130160640_add_unique_constraint_to_entries.rb]: I have a slight concern that this still leaves the possibility open to create a duplicate entry but with a different entry number.

File: db/migrate/20180130160640_add_unique_constraint_to_entries.rb (hunk 1 above)
Comment (arnau) [reply db/migrate/20180130160640_add_unique_constraint_to_entries.rb]: A bit late but, I'm missing `key` here. If `key` is part of the index, isn't unique enough to make `timestamp` relevant?

File: db/migrate/20180130160640_add_unique_constraint_to_entries.rb (hunk 2 above)
Comment (karlbaker02) [reply db/migrate/20180130160640_add_unique_constraint_to_entries.rb]: I'm not sure I understand what you mean here by a duplicate entry - could you elaborate please?

Comment (gidsg) [reply db/migrate/20180130160640_add_unique_constraint_to_entries.rb]: where the other fields are the same but the entry number has been mis-calculated. As we have seen examples of this.

File: db/migrate/20180130160640_add_unique_constraint_to_entries.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddUniqueConstraintToEntries < ActiveRecord::Migration[5.1]
+  def change
+    add_index(:entries, %i[hash_value entry_type entry_number spina_register_id key], unique: true, name: 'unique_entry_index')
```
Comment (danmitchell-) [inline db/migrate/20180130160640_add_unique_constraint_to_entries.rb]: Should be `register_id` now

**Pull request merged**